canlink/tracer.go

```diff
@@ -180,51 +114,67 @@ func (t *Tracer) Error() error {
 	return t.err.Err()
 }
 
-// fetchData fetches CAN frames from the socket and sends them over a buffered channel
-func (t *Tracer) fetchData(ctx context.Context) {
+// Handle listens to the frames in the broadcastChan and writes them to a file
+func (t *Tracer) Handle(broadcastChan chan TimestampedFrame, transmitChan chan TimestampedFrame) error {
+	file, err := createEmptyTraceFile(t.traceDir, t.busName, t.converter.GetFileExtention())
+	t.traceFile = file
+	if err != nil {
+		t.l.Info("cannot create trace file")
+		return errors.Wrap(err, "creating trace file")
+	}
 
-	timeFrame := TimestampedFrame{}
+	if err != nil {
+		return errors.Wrap(err, "close trace file")
+	}
 
-	for t.receiver.Receive() {
-		select {
-		case <-ctx.Done():
-			t.l.Info("context deadline exceeded")
-			return
-		case _, ok := <-t.frameCh:
-			if !ok {
-				t.l.Info("frame channel closed, exiting fetch routine")
-				return
-			}
-		default:
-			timeFrame.Frame = t.receiver.Frame()
-			timeFrame.Time = time.Now()
+	timeout := time.After(t.timeout)
 
-			if t.isRunning {
-				t.frameCh <- timeFrame
+	func() error {
+		for {
+			select {
+			case <-timeout:
+				t.l.Info("maximum trace time reached")
+				return nil
+			case receivedFrame := <-broadcastChan:
```

Comment:
great to see this section abstracted away from the Converters

@samparent97 This is what the `Handler` interface is for. A `Tracer` is an example of a `Handler` whose job is to CAN frames to a file. By using the Handler interface, we don't need to re-implement CAN message brokering